guide/src/class/numeric.md

```diff
@@ -0,0 +1,373 @@
+# Emulating numeric types
+
+At this point we have a `Number` class that we can't actually do any math on!
+
+Before proceeding, we should think about how we want to handle overflows. There are three obvious solutions:
+- We can have infinite precision just like Python's `int`. However that would be quite boring - we'd
+ be reinventing the wheel.
+- We can raise exceptions whenever `Number` overflows, but that makes the API painful to use.
+- We can wrap around the boundary of `i32`. This is the approach we'll take here. To do that we'll just forward to `i32`'s
+ `wrapping_*` methods. 
+
+### Fixing our constructor
+
+Let's address the first overflow, in `Number`'s constructor:
+
+```python
+from my_module import Number
+
+n = Number(1 << 1337)
+```
+
+```text
+Traceback (most recent call last):
+  File "example.py", line 3, in <module>
+    n = Number(1 << 1337)
+OverflowError: Python int too large to convert to C long
+```
+
+Instead of relying on the default [`FromPyObject`] extraction to parse arguments, we can specify our
+own extraction function, using the `#[pyo3(from_py_with = "...")]` attribute. Unfortunately PyO3
+ doesn't provide a way to wrap Python integers out of the box, but there's a Python C API function that does:
+
+```c
+unsigned long PyLong_AsUnsignedLongMask(PyObject *obj)
+```
```

Comment:
Should we add one? It might be nice to not encourage beginners to use `unsafe`.

Reply:
They can go through the python apis to do `x & 0xFFFFFFFF` before extracting. So strictly speaking unsafe is not necessary.

Maybe I should do that and make an appendix with the unsafe way later.